langchain/src/vectorstores/vectara.ts

```diff
@@ -41,30 +49,177 @@ export interface VectaraFile {
   lambda?: number;
-  // The number of sentences before/after the matching segment to add to the context.
+  // See Vectara Search API docs for more details on the following options: https://docs.vectara.com/docs/api-reference/search-apis/search
   contextConfig?: VectaraContextConfig;
+  mmrConfig?: MMRConfig;
 }
 
-/**
- * Interface for the context configuration used in Vectara API calls.
- */
-export interface VectaraContextConfig {
-  // The number of sentences before the matching segment to add. Default is 2.
-  sentencesBefore?: number;
-  // The number of sentences after the matching segment to add. Default is 2.
-  sentencesAfter?: number;
+const DEFAULT_FILTER: VectaraFilter = {
+  start: 0,
+  filter: "",
+  lambda: 0.0,
+  contextConfig: {
+    sentencesBefore: 2,
+    sentencesAfter: 2,
+    startTag: "<b>",
+    endTag: "</b>",
+  },
+  mmrConfig: {
+    enabled: false,
+    mmrTopK: 0,
+    diversityBias: 0.0,
+  },
+};
+
+interface SummaryResult {
+  documents: Document[];
+  scores: number[];
+  summary: string;
+}
+
+export interface VectaraRetrieverInput {
+  vectara: VectaraStore;
+  topK: number;
+  summaryConfig?: VectaraSummary;
+  callbacks?: Callbacks;
+  tags?: string[];
+  metadata?: Record<string, unknown>;
+  verbose?: boolean;
+}
+
+export class VectaraRetriever extends Runnable<string, Document[]> {
+  static lc_name() {
+    return "VectaraRetriever";
+  }
+
+  lc_namespace = ["langchain", "retrievers", "vectaraRetriever"];
+
+  callbacks?: Callbacks;
+
+  tags?: string[];
+
+  metadata?: Record<string, unknown>;
+
+  verbose?: boolean;
+
+  private vectara: VectaraStore;
+  
+  private topK: number;
+
+  private summaryConfig: VectaraSummary;
+
+  constructor(fields: VectaraRetrieverInput) {
+    super(fields);
+    this.callbacks = fields?.callbacks;
+    this.tags = fields?.tags ?? [];
+    this.metadata = fields?.metadata ?? {};
+    this.verbose = fields?.verbose ?? false;
+    this.vectara = fields.vectara;
+    this.topK = fields.topK ?? 10;
+    this.summaryConfig = fields.summaryConfig ?? { enabled: false, maxSummarizedResults: 0, responseLang: "eng" };
+  }
+
+  async invoke(input: string, options?: RunnableConfig): Promise<Document[]> {
+    return this.getRelevantDocuments(input, options);
+  }
+
+  
+  async getRelevantDocuments(
+    query: string,
+    config: VectaraFilter = DEFAULT_FILTER,
+    summary = false,
+  ): Promise<Document[]> {  
+    const parsedConfig = parseCallbackConfigArg(config);
+    const callbackManager_ = await CallbackManager.configure(
+      parsedConfig.callbacks,
+      this.callbacks,
+      parsedConfig.tags,
+      this.tags,
+      parsedConfig.metadata,
+      this.metadata,
+      { verbose: this.verbose }
+    );
+    const runManager = await callbackManager_?.handleRetrieverStart(
+      this.toJSON(),
+      query,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      parsedConfig.runName
+    );
```

Comment:
Handling run managers and this awful API are all unnecessary if you override `_getRelevantDocuments` instead of this class:

https://github.com/langchain-ai/langchainjs/blob/main/langchain-core/src/retrievers.ts#L70

Reply:
The issue is that since my getRelevantDocs has "summary" argument that does not fit with the signature of _getRelevantDocuments(). Is there an easy way around it?
This is why I did not inherit from BaseRetriever and had to copu/paste all this API stuff.